# Cleaning & Analyzing Employee Exit Surveys

In this project we're going to be working with exit surveys from employees of the Department of Education, Training and Employment(DETE) and the Technical and Further Education (TAFE) insitute in Queensland, Australia.

Our goal is to find out the following ;
- Are employees who only worked for the insitutes for a short period of time resigning due to some kind of dissatisfaction ? What about employees who have been there longer ?
- Are younger employees resigning due to some kind of dissatisfaction ? What about older employees ?

We'll combine the results from both surveys to answer these questions. Some of the columns we'll be working with from the DETE survey are;
- ID : an id used to identify the participant of the survey
- SeparationType : the reason why the person's employment ended
- Cease Date : the year or month the person's employment ended
- DETE Start Date : the year the person began employment with DETE

The columns we'll be working with from the TAFE survey are;
- Record ID : an id used to identify the participant of the survey 
- Reason for ceasing employment : the reason why the person's emlpoyment ended
- LengthofService Overall. Overall Length of Service at Insitute (in years) : length of the person's employment

In [32]:
# import numpy and pandas
import numpy as np
import pandas as pd

# read in DETE and TAFE csv
dete_survey = pd.read_csv('dete_survey.csv')
tafe_survey = pd.read_csv('tafe_survey.csv')


In [33]:
# dete info
dete_survey.info

<bound method DataFrame.info of       ID                        SeparationType  Cease Date DETE Start Date  \
0      1                 Ill Health Retirement     08/2012            1984   
1      2      Voluntary Early Retirement (VER)     08/2012      Not Stated   
2      3      Voluntary Early Retirement (VER)     05/2012            2011   
3      4             Resignation-Other reasons     05/2012            2005   
4      5                        Age Retirement     05/2012            1970   
5      6             Resignation-Other reasons     05/2012            1994   
6      7                        Age Retirement     05/2012            1972   
7      8                        Age Retirement     05/2012            1988   
8      9             Resignation-Other reasons     07/2012            2009   
9     10            Resignation-Other employer        2012            1997   
10    11                        Age Retirement        2012            1999   
11    12  Resignation-Move overs

In [34]:
# first 5 of dete
dete_survey.head(5)

ID                    SeparationType Cease Date DETE Start Date  \
0   1             Ill Health Retirement    08/2012            1984   
1   2  Voluntary Early Retirement (VER)    08/2012      Not Stated   
2   3  Voluntary Early Retirement (VER)    05/2012            2011   
3   4         Resignation-Other reasons    05/2012            2005   
4   5                    Age Retirement    05/2012            1970   

  Role Start Date                                      Position  \
0            2004                                Public Servant   
1      Not Stated                                Public Servant   
2            2011                               Schools Officer   
3            2006                                       Teacher   
4            1989  Head of Curriculum/Head of Special Education   

  Classification              Region                      Business Unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-AO7      Central Office  Corporate Strategy and Peformance   
2            NaN      Central Office               Education Queensland   
3        Primary  Central Queensland                                NaN   
4            NaN          South East                                NaN   

     Employment Status  ...   Kept informed  Wellness programs  \
0  Permanent Full-time  ...               N                  N   
1  Permanent Full-time  ...               N                  N   
2  Permanent Full-time  ...               N                  N   
3  Permanent Full-time  ...               A                  N   
4  Permanent Full-time  ...               N                  A   

   Health & Safety  Gender          Age  Aboriginal  Torres Strait  South Sea  \
0                N    Male        56-60         NaN            NaN        NaN   
1                N    Male        56-60         NaN            NaN        NaN   
2                N    Male  61 or older         NaN            NaN        NaN   
3                A  Female        36-40         NaN            NaN        NaN   
4                M  Female  61 or older         NaN            NaN        NaN   

   Disability  NESB  
0         NaN   Yes  
1         NaN   NaN  
2         NaN   NaN  
3         NaN   NaN  
4         NaN   NaN  

[5 rows x 56 columns]

In [35]:
# tafe info
tafe_survey.info

<bound method DataFrame.info of         Record ID                              Institute  \
0    6.341330e+17  Southern Queensland Institute of TAFE   
1    6.341337e+17            Mount Isa Institute of TAFE   
2    6.341388e+17            Mount Isa Institute of TAFE   
3    6.341399e+17            Mount Isa Institute of TAFE   
4    6.341466e+17  Southern Queensland Institute of TAFE   
5    6.341475e+17  Southern Queensland Institute of TAFE   
6    6.341520e+17         Barrier Reef Institute of TAFE   
7    6.341537e+17  Southern Queensland Institute of TAFE   
8    6.341579e+17  Southern Queensland Institute of TAFE   
9    6.341588e+17         Barrier Reef Institute of TAFE   
10   6.341588e+17         Barrier Reef Institute of TAFE   
11   6.341719e+17         Barrier Reef Institute of TAFE   
12   6.341719e+17         Barrier Reef Institute of TAFE   
13   6.341725e+17         Barrier Reef Institute of TAFE   
14   6.341726e+17   Central Queensland Institute of TAFE   
15   6.3

In [36]:
# first 5 of tafe 
tafe_survey.head(5)

Record ID                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   WorkArea  CESSATION YEAR Reason for ceasing employment  \
0  Non-Delivery (corporate)          2010.0              Contract Expired   
1  Non-Delivery (corporate)          2010.0                    Retirement   
2       Delivery (teaching)          2010.0                    Retirement   
3  Non-Delivery (corporate)          2010.0                   Resignation   
4       Delivery (teaching)          2010.0                   Resignation   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

                                     ...                                     \
0                                    ...                                      
1                                    ...                                      
2                                    ...                                      
3                                    ...                                      
4                                    ...                                      

  Workplace. Topic:Does your workplace promote a work culture free from all forms of unlawful discrimination?  \
0                                                Yes                                                            
1                                                Yes                                                            
2                                                Yes                                                            
3                                                Yes                                                            
4                                                Yes                                                            

  Workplace. Topic:Does your workplace promote and practice the principles of employment equity?  \
0                                                Yes                                               
1                                                Yes                                               
2                                                Yes                                               
3                                                Yes                                               
4                                                Yes  

After reviewing the first 5 rows of each dataframe we can see that in the DETE survey there is "Not Stated" values but they are not represented as NaN values. Some of the columns are the same in each data set but they're named differently, there is also a handful of columns that we don't need to complete our anaylsis. 

In [37]:
# sum of all null values in each column in dete survey
dete_survey.isnull().sum()

ID                                       0
SeparationType                           0
Cease Date                               0
DETE Start Date                          0
Role Start Date                          0
Position                                 5
Classification                         367
Region                                   0
Business Unit                          696
Employment Status                        5
Career move to public sector             0
Career move to private sector            0
Interpersonal conflicts                  0
Job dissatisfaction                      0
Dissatisfaction with the department      0
Physical work environment                0
Lack of recognition                      0
Lack of job security                     0
Work location                            0
Employment conditions                    0
Maternity/family                         0
Relocation                               0
Study/Travel                             0
Ill Health 

Above we can see that business unit, aboriginal, torres strait, south sea, disability, nesb have over 600 null values. Additionally these are potential drop candidates since the data within these columns most likely won't help us in answering the questions we stated at the beginning of the project.

In [38]:
# sum of all null values in columns of tafe survey
tafe_survey.isnull().sum()

Record ID                                                                                                                                                          0
Institute                                                                                                                                                          0
WorkArea                                                                                                                                                           0
CESSATION YEAR                                                                                                                                                     7
Reason for ceasing employment                                                                                                                                      1
Contributing Factors. Career Move - Public Sector                                                                                                                265
Contributi

There is many different Factors columns within the tafe survey, all which have well over 200 missing values.

Next we are going to read the DETE survey again except this time we'll read the 'Not Stated' values in as NaN values by setting the na_values parameter to Not Stated.

In [39]:
dete_survey = pd.read_csv('dete_survey.csv', na_values = 'Not Stated')

Next lets drop some columns from each dataframe that we wont be using first starting with the dete survey in which we'll drop columns 28 through 48.

In [40]:
dete_survey_updated = dete_survey.drop(dete_survey.columns[28:49], axis=1)

We'll do the same with the tafe survey but for columns 17 through 65.

In [41]:
tafe_survey_updated = tafe_survey.drop(tafe_survey.columns[17:66], axis=1)

Now that we've dropped columns we won't be using, we can focus on column names since each dataframe contains many of the same columns but they're named differently. We'll start with the dete survey and make the capitalization lowercase, remove any trailing whitespace from the end of the strings, and replace spaces with underscores.

In [42]:
dete_survey_updated.columns = dete_survey_updated.columns.str.lower().str.strip().str.replace(' ','_')

Next we'll use the dataframe.rename() method to update some of the columns in the tafe survey.

In [48]:
tafe_survey_updated = tafe_survey_updated.rename(columns={'Record ID':'id',
                                                          'CESSATION YEAR': 'cease_date',
                                                          'Reason for ceasing employment':'separationtype',
                                                          'Gender. What is your Gender':'gender',
                                                          'CurrentAge. Current Age':'age',
                                                          'Emplyoment Type. Employment Type':'emplyoment_status',
                                                          'Classification. Classification':'position',
                                                          'LengthofServiceOverall. Overall Length of Service at Institute (in years)':'institute_service',
                                                          'LengthofServiceCurrent. Length of Service at current workplace (in years)':'role_service'})

In [46]:
dete_survey_updated.head(2)

id                    separationtype cease_date  dete_start_date  \
0   1             Ill Health Retirement    08/2012           1984.0   
1   2  Voluntary Early Retirement (VER)    08/2012              NaN   

   role_start_date        position classification          region  \
0           2004.0  Public Servant        A01-A04  Central Office   
1              NaN  Public Servant        AO5-AO7  Central Office   

                       business_unit    employment_status  ...   \
0  Corporate Strategy and Peformance  Permanent Full-time  ...    
1  Corporate Strategy and Peformance  Permanent Full-time  ...    

   work_life_balance  workload  none_of_the_above  gender    age  aboriginal  \
0              False     False               True    Male  56-60         NaN   
1              False     False              False    Male  56-60         NaN   

   torres_strait  south_sea  disability  nesb  
0            NaN        NaN         NaN   Yes  
1            NaN        NaN         NaN   NaN  

[2 rows x 35 columns]

In [49]:
tafe_survey_updated.head(2)

id                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   

                   WorkArea  cease_date    separationtype  \
0  Non-Delivery (corporate)      2010.0  Contract Expired   
1  Non-Delivery (corporate)      2010.0        Retirement   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   

      ...      Contributing Factors. Study Contributing Factors. Travel  \
0     ...                              NaN                          NaN   
1     ...                                -                       Travel   

  Contributing Factors. Other Contributing Factors. NONE  \
0                         NaN                        NaN   
1                           -                          -   

  Gender. What is your Gender?     age Employment Type. Employment Type  \
0                       Female  26  30              Temporary Full-time   
1                          NaN     NaN                              NaN   

              position institute_service role_service  
0  Administration (AO)               1-2          1-2  
1                  NaN               NaN          NaN  

[2 rows x 23 columns]

Above we can confirm that our changes were executed.

In [53]:
# using value_counts() method in seperationtype column in both datasets
print(dete_survey_updated['separationtype'].value_counts())
print(tafe_survey_updated['separationtype'].value_counts())

Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15
Name: separationtype, dtype: int64
Resignation                 340
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23
Name: separationtype, dtype: int64


In [54]:
# use regex pattern to find string values containing Resignation/resignation
pattern = r'[Rr]esignation'

# use .loc and str.contains along with pattern to match strings in seperationtype
dete_resignations = dete_survey_updated.loc[dete_survey_updated['separationtype'].str.contains(pattern)]
tafe_resignations = tafe_survey_updated.loc[tafe_survey_updated['separationtype'].str.contains(pattern,na=False)]

# we'll use dataframe.copy() method to avoid the SettingWithCopy Warning
dete_resignations = dete_resignations.copy()
tafe_resignations = tafe_resignations.copy()

The code above allows us to gather all of the data that has to do with resignation.

Next we're going to clean the cease_date column within dete_resignations. We'll use the value_counts() method to view the unique values in the cease_date column, use vectorized string methods to extract the year, and then use the series.astype() method to convert the type to a float.

In [55]:
dete_resignations['cease_date'].value_counts()

2012       126
2013        74
01/2014     22
12/2013     17
06/2013     14
09/2013     11
07/2013      9
11/2013      9
10/2013      6
08/2013      4
05/2012      2
05/2013      2
07/2006      1
2010         1
07/2012      1
09/2010      1
Name: cease_date, dtype: int64

In [56]:
# regex pattern to extract the dates with str.extract(regex pattern, expand=True)
pattern = r'([2][0-9]{3})'
dete_resignations['cease_date'] = dete_resignations['cease_date'].str.extract(pattern, expand=True)
dete_resignations['cease_date'] = dete_resignations['cease_date'].astype(float)

In [57]:
dete_resignations['cease_date'].value_counts()

2013.0    146
2012.0    129
2014.0     22
2010.0      2
2006.0      1
Name: cease_date, dtype: int64

In [59]:
tafe_resignations['cease_date'].value_counts()

2011.0    116
2012.0     94
2010.0     68
2013.0     55
2009.0      2
Name: cease_date, dtype: int64

The span of the years between both dataframes are not completely the same but they're close enough ranges that we do not need to drop any additional data.

The TAFE dataframe already contains a service column which is the amount of years that employee worked at that institution, we renamed the column institute_service. The DETE dataframe does not have this column so we'll create the column for the dataframe and since we have the data for dete_start_date and cease_date we'll subtract the two and fill our name institute_service column with that data.

In [62]:
dete_resignations['institute_service'] = dete_resignations['cease_date'] - dete_resignations['dete_start_date']

In [63]:
dete_resignations.head(3)

id             separationtype  cease_date  dete_start_date  \
3   4  Resignation-Other reasons      2012.0           2005.0   
5   6  Resignation-Other reasons      2012.0           1994.0   
8   9  Resignation-Other reasons      2012.0           2009.0   

   role_start_date          position classification              region  \
3           2006.0           Teacher        Primary  Central Queensland   
5           1997.0  Guidance Officer            NaN      Central Office   
8           2009.0           Teacher      Secondary    North Queensland   

          business_unit    employment_status        ...          workload  \
3                   NaN  Permanent Full-time        ...             False   
5  Education Queensland  Permanent Full-time        ...             False   
8                   NaN  Permanent Full-time        ...             False   

   none_of_the_above  gender    age  aboriginal  torres_strait  south_sea  \
3              False  Female  36-40         NaN            NaN        NaN   
5              False  Female  41-45         NaN            NaN        NaN   
8              False  Female  31-35         NaN            NaN        NaN   

   disability  nesb  institute_service  
3         NaN   NaN                7.0  
5         NaN   NaN               18.0  
8         NaN   NaN                3.0  

[3 rows x 36 columns]

We can see above that our new column is the last column of the data frame with its new values.

Next we're going to created a new column called dissatisfied that will return True if they left becuase of dissatisfaction, False if they did not leave due to dissatisfaction and NaN so we don't lose data. We'll do this by creating a function that runs through all the dissatisfaction values. 

In [64]:
def update_values(value):
    if pd.isnull(value):
        return np.NaN
    elif value == '-':
        return False
    else:
        return True

In [70]:
# creating dissatisfaction column
dete_resignations['dissatisfaction'] = dete_resignations[['job_dissatisfaction',
       'dissatisfaction_with_the_department', 'physical_work_environment',
       'lack_of_recognition', 'lack_of_job_security', 'work_location',
       'employment_conditions', 'work_life_balance','workload']].any(1, skipna=False)

# using .copy() to avoid SettingWithCopy Warning

dete_resignations_up = dete_resignations.copy()
dete_resignations_up['dissatisfaction'].value_counts()

False    162
True     149
Name: dissatisfaction, dtype: int64

In [71]:
tafe_columns = ['Contributing Factors. Dissatisfaction','Contributing Factors. Job Dissatisfaction']

# use applymap() method to apply update_values function to tafe
tafe_resignations['dissatistfaction'] = tafe_resignations[tafe_columns].applymap(update_values).any(1,skipna=False)

# using .copy() to avoid SettingWithCopy Warning
tafe_resignations_up = tafe_resignations.copy()

In [72]:
tafe_resignations['dissatistfaction'].value_counts()

False    241
True      91
Name: dissatistfaction, dtype: int64

We can see that in both surveys more employees did not quit due to dissatistfaction, in the dete survey however it was much closer than tafe with 162 False values and 149 true values.

Next we're going to work on combining our datasets, first we'll add a column to each dataframe name institue so we can distinguish between DETE and TAFE, we'll combine them into a dataframe named combined, and use the dataframe.dropna() method to drop any columns with less than 500 non null values by making sure we set the thresh paramater. The final result of this will be stored in combined_update.

In [73]:
# creating institute column for each dataframe
dete_resignations_up['institute'] = 'DETE'
tafe_resignations_up['institute'] = 'TAFE'

# combining both with pd.concat
combined = pd.concat([dete_resignations_up, tafe_resignations_up], ignore_index=True)

In [74]:
combined.notnull().sum().value_counts()

311    19
332    13
290     3
651     3
340     2
596     1
563     1
3       1
7       1
8       1
9       1
271     1
598     1
635     1
283     1
32      1
161     1
302     1
265     1
307     1
0       1
dtype: int64